### ISSUE TITLE
kube-apiserver-arg has random order after each mutate

### ISSUE THREAD

**Issue by hansbogert** [rancher/rancher#43274]
Everytime the admission mutating webhook finishes, it changes the order of `kube-apiserver-arg`

This causes the behaviour in https://github.com/rancher/terraform-provider-rancher2/issues/1241 

The obvious thing to do (i think), is to sort the `parsed` list here: https://github.com/rancher/webhook/blob/e401287e9f7f36f1a3d718c68918b0a3b3227147/pkg/resources/provisioning.cattle.io/v1/cluster/mutator.go#L289

This would make the list order non-random and eventually consistent such that the Terraform state will become stable as well.

/update
As in https://github.com/rancher/terraform-provider-rancher2/issues/1241, I could only reproduce this if you have a  `default_pod_security_admission_configuration_template_name` set in the terraform `rancher2_cluster_v2` resource and you include other apiserver-args.

**Comment by hansbogert** [comment]
I can confirm that this would alleviate, or possibly solve the problem:

```patch
@@ -286,6 +305,11 @@ func setKubeAPIServerArg(arg map[string]string, cluster *v1.Cluster) {
        if cluster.Spec.RKEConfig.MachineGlobalConfig.Data == nil {
                cluster.Spec.RKEConfig.MachineGlobalConfig.Data = make(map[string]interface{})
        }
+
+        sort.Slice(parsed, func(i,j int) bool {
+           return fmt.Sprintf("%v",parsed[i]) < fmt.Sprintf("%v", parsed[j]);
+        })
+
        cluster.Spec.RKEConfig.MachineGlobalConfig.Data["kube-apiserver-arg"] = parsed
 }
```

 though I'm reluctant to say that sorting here is the best way to make the problem go away on terraform level; I can imagine you could solve this problem in terraform, in the provider, as well.

**Comment by atsai1220** [comment]
We're also running into this issue as we're using ArgoCD to sync our `cluster.provisioning` objects and the `kube-apiserver-arg` list keeps getting into a random order.

**Comment by DrackThor** [comment]
We run into a similar problem that might result from the same cause.
When I try to provide an additional admission-control-config-file, it just get's dropped by Rancher.
In the HCL I have somehting like this..

```hcl
machine_global_config = yamlencode({
      ...
      kube-apiserver-arg  = [
        ...
        "admission-control-config-file=/etc/rancher/rke2/admission-config.yaml",
```

The values show up correctly in the terraform plan.
The created config in Rancher however looks like this -> only the default PSACT admission config is present here.

```yaml
machineGlobalConfig:
      ...
      kube-apiserver-arg:
        ...
        - enable-admission-plugins=AlwaysPullImages,NodeRestriction,EventRateLimit
        - admission-control-config-file=/etc/rancher/rke2/config/rancher-psact.yaml
        ...
```
Even when I try to change the `admission-control-config-file` from `/etc/rancher/rke2/config/rancher-psact.yaml` to `/etc/rancher/rke2/admission-config.yaml`, or just add an additional line `"admission-control-config-file=/etc/rancher/rke2/admission-config.yaml"` in the Rancher UI, it gets reverted to the original `admission-control-config-file=/etc/rancher/rke2/config/rancher-psact.yaml`.

Long story short, we're not able to provide an additional or override the existing `admission-control-config-file` arg.

**Comment by pwurbs** [comment]
I agree that this is a severe issue. We are encouraged to configure the cluster in a safe way (satisfy CIS scan), but Rancher blocks us to set individual AdmissionConfiguration. I always end up having to times the admission-control-config-file argument in the kube-apiserver pod yaml:
```
--admission-control-config-file=/etc/rancher/rke2/rke2-pss.yaml'
--admission-control-config-file=/etc/rancher/rke2/admission-control.yaml 
```
This is dangerous, because now it depends, which file is used. If the 1st one is used w/o config of EventRateLimit, then the node doesn't start.
I could narrow it down a bit more and exclude any own adaptions:
- Created a fresh RKE2 downstream cluster (1.28) using the Rancher UI, activated the CIS profile, configured to use the rancher-restricted PSACT
- Cluster YAML looks good
- Even the automatically generated RKE2 cluster config at /etc/rancher/rke2/config.yaml.d/50-rancher.yaml looks good, only contains   "kube-apiserver-arg": [    "admission-control-config-file=/etc/rancher/rke2/config/rancher-psact.yaml" ]
- But the started kube-apiserver pod contains the admission-control-config-file argument twice:
```
    - args:
        - '--admission-control-config-file=/etc/rancher/rke2/rke2-pss.yaml'
        - '--admission-control-config-file=/etc/rancher/rke2/config/rancher-psact.yaml'
```
Again, the rke2-pss.yaml has been added. 
I have the suspicion, that the unwished addition is due to the activated CIS profile. 
We don't see this behavior in our Rancher Clusters. So, I think the automatic addition is done by Rancher and not by RKE2.

**Comment by mandronic0** [comment]
SURE-9741

**Comment by jiaqiluo** [comment]
The issue is reproduced on Rancher v2.11-head, and I believe it’s reproducible on other versions as well.

Using Terraform-rancher-provider is not requried, and it can be reproduced via Rancher UI. 


**Steps:**
- Create a custom K3s cluster with setting the PSACT to any value, such as "rancher-privilege", and some other kube-apiserver arguments, such as the following
```
      kube-apiserver-arg:
        - >-
          admission-control-config-file=/etc/rancher/k3s/config/rancher-psact.yaml
        - feature-gates=AllAlpha=false
        - tls-min-version=VersionTLS13
        - enable-admission-plugins=AlwaysPullImages,NodeRestriction
        - audit-log-format=json
        - audit-log-maxage=30
        - audit-log-maxbackup=10
        - audit-log-maxsize=100
```
- Click the save button to create the cluster 
- Click "Edit YAML" on the cluster, do not make any chage, but check the values under the  `kube-apiserver-arg` 
- Repeat the last step multiple times 

**Results:**

The order of args in the `kube-apiserver-arg` may change each randomly.

**Comment by susesgartner** [comment]
Ticket #43274 - Test Results - ✅
---
 
| Scenario | Test Case | Result |
|---|---|---|
| 1. | Verify kube args order after mutating | PASS |
 
---
Verified with (HA Helm or Docker) on Rancher `v2.12-d1c101cf9a5c44f21540900a04d9343535d2fcfd-head`:
---
 
Scenario 1: Verify kube args order after mutating
---
1. Provision a k3s custom cluster with the following kube-apiserver args:
      kube-apiserver-arg:
        - >-
          admission-control-config-file=/etc/rancher/k3s/config/rancher-psact.yaml
        - feature-gates=AllAlpha=false
        - tls-min-version=VersionTLS13
        - enable-admission-plugins=AlwaysPullImages,NodeRestriction
        - audit-log-format=json
        - audit-log-maxage=30
        - audit-log-maxbackup=10
        - audit-log-maxsize=100
2. Verify that the args remain in the correct order
3. Update a cluster setting.
4. Verify that the args remain in the correct order
 
Results: The args remain in the correct order
